[Internal Review] Add per-host configs

k4r1 opened the pull request:
**Notice**
> This PR is for internal review
> Once we've checked it over, we'll take the branch and initiate a PR into the upstream repo if we feel it is relevant

---

# Per Host Configs

This PR introduces a new repeatable flag to the CLI called `--config-set`. A config-set is a per-host configuration of the override values for the proxy, namely:
- name (which aws service to invoke, for example `execute-api`)
- host
- role-arn
- region

Using these flags, you can configure the proxy to use a different configuration based on which host is being requested. For example:

```
./aws-sigv4-proxy \
    --config-set '{name: execute-api, host: x.myapi.com, region: eu-west-1, role-arn: arn:aws:iam::123:role/invoke-myapi-role}' \
    --config-set '{name: execute-api, host: x.otherapi.com, region: ap-southeast-1, role-arn: arn:aws:iam::456:role/invoke-otherapi-role}' 
```

If the domain does not match any of the config sets, the proxy will default to its original (existing) behaviour.

![image](https://user-images.githubusercontent.com/16408267/106113454-7833b300-6189-11eb-83b2-15b9b8376b48.png)

k4r1 committed:
Initial implementation idea for per-host configs

k4r1 committed:
Add test for per-host config

lucas-sonnabend reviewed (commented):
golang stuff looks okay to me, but I'm not an expert

sgrad commented on handler/config_set.go:
```diff
@@ -0,0 +1,10 @@
+
+// ConfigSet contains overrides for individual hosts
+type ConfigSet struct {
+	Name string
```
It's good practice to label all fields to avoid struct name changes breaking existing configs

sgrad commented on main.go:
```diff
@@ -38,45 +39,79 @@ var (
-		log.SetLevel(log.DebugLevel)
-	}
 
+func getSigner(roleArn string) *v4.Signer {
```
NIT: no need to use value here, you're copying the string when calling the function, just pass a string pointer

sgrad commented on main.go:
```diff
@@ -38,45 +39,79 @@ var (
+		log.SetLevel(log.DebugLevel)
+	}
+
+	signer := getSigner(*roleArn)
```
change the func to take a pointer and pass without *

sgrad commented on main.go:
```diff
@@ -38,45 +39,79 @@ var (
+	for _, configSetYaml := range *configSets {
+		configSet := handler.ConfigSet{}
+
+		err := yaml.Unmarshal([]byte(configSetYaml), &configSet)
```
NIT: you can collapse this into one instruction:
```
if err := call(); err ~= nil {...}
```

sgrad commented on main.go:
```diff
@@ -38,45 +39,79 @@ var (
+			"Name": configSet.Name,
+			"Region": configSet.Region,
+			"RoleArn": configSet.RoleArn,
+		}).Infof("Adding config for host")
```
If you're not formatting the string there's no need for `Infof` just use `Info`

sgrad commented on main.go:
```diff
@@ -38,45 +39,79 @@ var (
 	log.WithFields(log.Fields{"port": *port}).Infof("Listening on %s", *port)
 
+	clients := map[string]handler.Client{
+		"default": &handler.ProxyClient{
```
This is not ideal, the handler is always expecting the default to be set, and it's manually set in the main func. Ideally, you'd use a separate constructor parameter, or (since the whole aws implementation is not that well structured) just add an error check in the handler when getting the `default` val so it breaks if anyone changes the main function call

k4r1 replied:
Added the runtime check

sgrad commented on main.go:
```diff
@@ -38,45 +39,79 @@ var (
+			RegionOverride: configSet.Region,
+		}
+	}
+
```
Run gofmt on the code please

k4r1 replied:
The problem is that the existing code is not formatted using `gofmt`, so running it changes basically all of the existing code

k4r1 committed:
Cleanup based on feedback
- Use shorthand for determining yaml unmarshall error
- Use string pointer (instead of copying string) for getSigner method
- Replace unused .Infof with .Info

k4r1 committed:
Map yaml to struct explicitly with tags

k4r1 committed:
Add runtime check for existence of default client
